src/www/firewall_rules_edit.php

```diff
@@ -40,8 +40,9 @@
 if ($ostypes == null) {
     $ostypes = array();
 }
+$ipprotocols = array('inet', 'inet6', 'inet46');
```

Comment:
basically we have two options: validate *all* pages that allow such input consistently

or leave it unvalidated having XSS issues worst case (if pages are not protected which they normally are) but then again the rule generator underneath should be clever enough to discard garbage so that your rules are not tainted. I'm fairly certain you can't inject rules anymore, which is the only way this would be going. And it would also argue you'd inject a rule in order to be able to change rules what you already can submit in the first place.

validation of drop downs in general is a matter of debate and always happens deliberately so the basic question is what is the thread model?

Reply:
I'm not sure. I'm not a security specialist in that sense. I do think that if one follows that line, the argument becomes "we prevent a malicious actor entering unwanted data as best we can, but beyond that, junk is OK in any field because it gets blocked at the point of use".  

Which is, I suppose valid but seems pretty unsatisfactory. Do we validate that ports are inputted as 1-65535 only, as well as check at the point we generate pf.conf? What about IP addresses, would we let people enter ";rm *" or whatever as an IP address and accept that into $config, on the basis that it's fine because it'll be caught when IP-based config is built and fields are validated? Or XML characters? Why have $input_errors[] anyway if we catch on use? 

I just... I'm not sure the threat model or how a secuirity specialist would see it. I just feel profoundly disturbed that any form would accept arbitrary text and insert it into the config, whether or not we believe it'll be caught later and rejected upon actual use.

Surely it makes sense to fix if a fix is offered, even if its just one page at present and even if the legacy code will at some time be superseded.

Reply:
In these cases a top down would be best. Set a strategy for invalid input. If case 2 and 4 are bad, maybe even case 1 why not throw an input error right away?